File: cpp/74_search_a_2d_matrix.cpp
#include <vector>

using namespace std;

class Solution {
 public:
  bool searchMatrix(vector<vector<int>>& matrix, int target) {
    int m = matrix.size();
    int n = matrix[0].size();
    int low = 0, high = m * n - 1;
    while (low <= high) {
      int mid = low + (high - low) / 2;
      int val = matrix[mid / n][mid % n];
      if (val < target) {
        low = mid + 1;
      } else if (val > target) {
        high = mid - 1;
      } else {
        return true;
      }
    }
    return false;
  }
};

File: rust/74_search_a_2d_matrix.rs
use std::cmp::Ordering;

impl Solution {
    pub fn search_matrix(matrix: Vec<Vec<i32>>, target: i32) -> bool {
        let (m, n) = (matrix.len(), matrix[0].len());
        let (mut low, mut high) = (0i32, (m * n) as i32 - 1);

        while low <= high {
            let mid = low + (high - low) / 2;
            let umid = mid as usize;
            let val = matrix[umid / n][umid % n];
            match val.cmp(&target) {
                Ordering::Equal => return true,
                Ordering::Less => low = mid + 1,
                Ordering::Greater => high = mid - 1,
            }
        }

        false
    }
}
